flytekit/types/iterator/json_iterator.py

```diff
@@ -0,0 +1,112 @@
+from pathlib import Path
+from typing import Any, Dict, Iterator, List, Type, Union
+
+import jsonlines
+from typing_extensions import TypeAlias
+
+from flytekit import FlyteContext, Literal, LiteralType
+from flytekit.core.type_engine import (
+    TypeEngine,
+    TypeTransformer,
+    TypeTransformerFailedError,
+)
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Scalar
+
+JSONCollection: TypeAlias = Union[Dict[str, Any], List[Any]]
+JSONScalar: TypeAlias = Union[bool, float, int, str]
+JSON: TypeAlias = Union[JSONCollection, JSONScalar]
+
+
+class JSONIterator:
+    def __init__(self, reader: jsonlines.Reader):
+        self._reader = reader
+        self._reader_iter = reader.iter()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return next(self._reader_iter)
+        except StopIteration:
+            self._reader.close()
+            raise StopIteration("File handler is exhausted")
+
+
+class JSONIteratorTransformer(TypeTransformer[Iterator[JSON]]):
```

Comment:
qq: why can't we use the existing IteratorTransformer? It should be able to serialize `Iterator[JSON]` as long as we have a JSON transformer.
https://github.com/flyteorg/flytekit/blob/644404d1840406f587a6368093a71acaa20626d1/flytekit/types/iterator/iterator.py#L56

Reply:
@pingsutw the iterator might have to be a little different from the usual Iterator transformer that's already present. 

* JSON streaming is necessary to reduce the memory footprint (the batch API only supports `.jsonl` files up to 100 MB in size, so it doesn't really help in this case, but that's the general idea.)
* the literal representation should be a JSONL file with all the JSON objects rather than a list (JSONL is a standard format for large datasets)

makes sense?

Reply:
Got it. Thanks for the explanation.